echopype/calibrate/cal_params.py

```diff
@@ -1,155 +1,249 @@
-        A subset of Vendor_specific that contains only the channels specified for calibration
-    user_cal_dict : dict
-        A dictionary containing user-defined calibration parameters.
-        The user-defined calibration parameters will overwrite values in the data file.
-
-    Returns
-    -------
-    A dictionary containing the calibration parameters for the EK echosounders
+    sonar_type : str
+        type of sonar, one of "EK60", "EK80", or "AZFP"
+    user_dict : dict
+        a dict containing user input calibration parameters as {parameter name: parameter value}
+        parameter value has to be a scalar (int or float) or an xr.DataArray
+        if parameter value is an xr.DataArray, it has to either have 'channel' as a coorindate
+        or have both 'cal_channel_id' and 'cal_frequency' as coordinates
+
+    channel : list or xr.DataArray
+        a list of channels to be calibrated
+        for EK80 data, this list has to corresponds with the subset of channels
+        selected based on waveform_mode and encode_mode
     """
-    out_dict = dict.fromkeys(CAL_PARAMS["EK"])
-    if user_cal_dict is None:
-        user_cal_dict = {}
+    # Check sonar type
+    if sonar_type not in ["EK60", "EK80", "AZFP"]:
+        raise ValueError("'sonar_type' has to be one of: 'EK60', 'EK80', or 'AZFP'")
 
-    # Params from the Beam group
-    params_from_beam = [
-        "angle_offset_alongship",
-        "angle_offset_athwartship",
-        "beamwidth_twoway_alongship",
-        "beamwidth_twoway_athwartship",
-        "equivalent_beam_angle",
-    ]
-    for p in params_from_beam:
-        # substitute if p not in user input
-        out_dict[p] = user_cal_dict.get(p, beam[p])
-
-    # Params from the Vendor_specific group
-    params_from_vend = ["sa_correction", "gain_correction"]
-    for p in params_from_vend:
-        # substitute if p not in user input
-        out_dict[p] = user_cal_dict.get(p, get_vend_cal_params_power(beam=beam, vend=vend, param=p))
+    # Make channel a sorted list
+    if not isinstance(channel, (list, xr.DataArray)):
+        raise ValueError("'channel' has to be a list or an xr.DataArray")
+    else:
+        if isinstance(channel, xr.DataArray):
+            channel_sorted = sorted(channel.data)
+        else:
+            channel_sorted = sorted(channel)
+
+    # Screen parameters: only retain those defined in CAL_PARAMS
+    #  -- transform params in scalar or list to xr.DataArray
+    #  -- directly pass through those that are xr.DataArray
+    out_dict = dict.fromkeys(CAL_PARAMS[sonar_type])
+    for p_name, p_val in user_dict.items():
+        if p_name in out_dict:
+            # if p_val an xr.DataArray, check existence and coordinates
+            if isinstance(p_val, xr.DataArray):
+                # if 'channel' is a coordinate, it has to match that of the data
+                if "channel" in p_val.coords:
+                    if not (sorted(p_val.coords["channel"].data) == channel_sorted):
```

Comment:
```suggestion
                    if not (sorted(p_val.coords["channel"].values) == channel_sorted):
```
See my comment on https://github.com/OSOceanAcoustics/echopype/pull/955/files#r1131764279... Maybe using values is better here.